tests/components/androidtv/test_media_player.py

```diff
@@ -1,232 +1,249 @@
 """The tests for the androidtv platform."""
 import logging
-from socket import error as socket_error
-import unittest
-from unittest.mock import patch
 
+from homeassistant.setup import async_setup_component
 from homeassistant.components.androidtv.media_player import (
-    AndroidTVDevice,
-    FireTVDevice,
-    setup,
+    ANDROIDTV_DOMAIN,
+    CONF_ADB_SERVER_IP,
+)
+from homeassistant.components.media_player.const import DOMAIN
+from homeassistant.const import (
+    CONF_DEVICE_CLASS,
+    CONF_HOST,
+    CONF_NAME,
+    CONF_PLATFORM,
+    STATE_IDLE,
+    STATE_OFF,
+    STATE_UNAVAILABLE,
 )
 
+from . import patchers
+
+
+# Android TV device with Python ADB implementation
+CONFIG_ANDROIDTV_PYTHON_ADB = {
+    DOMAIN: {
+        CONF_PLATFORM: ANDROIDTV_DOMAIN,
+        CONF_HOST: "127.0.0.1",
+        CONF_NAME: "Android TV",
+    }
+}
+
+# Android TV device with ADB server
+CONFIG_ANDROIDTV_ADB_SERVER = {
+    DOMAIN: {
+        CONF_PLATFORM: ANDROIDTV_DOMAIN,
+        CONF_HOST: "127.0.0.1",
+        CONF_NAME: "Android TV",
+        CONF_ADB_SERVER_IP: "127.0.0.1",
+    }
+}
+
+# Fire TV device with Python ADB implementation
+CONFIG_FIRETV_PYTHON_ADB = {
+    DOMAIN: {
+        CONF_PLATFORM: ANDROIDTV_DOMAIN,
+        CONF_HOST: "127.0.0.1",
+        CONF_NAME: "Fire TV",
+        CONF_DEVICE_CLASS: "firetv",
+    }
+}
+
+# Fire TV device with ADB server
+CONFIG_FIRETV_ADB_SERVER = {
+    DOMAIN: {
+        CONF_PLATFORM: ANDROIDTV_DOMAIN,
+        CONF_HOST: "127.0.0.1",
+        CONF_NAME: "Fire TV",
+        CONF_DEVICE_CLASS: "firetv",
+        CONF_ADB_SERVER_IP: "127.0.0.1",
+    }
+}
+
+
+async def _test_reconnect(hass, caplog, config):
```

Comment:
I think this is ok, but another approach that I would prefer, is to try to extract the set up and patching code into pytest fixtures and keep the assertions in the main test functions. It's easier to follow what each test should do, that way.

Reply:
I'm not familiar with pytest fixtures. 

Also, it's basically the same two tests repeated four times each (Python ADB vs. ADB server, Android TV device vs. Fire TV device). I thought it was more clear to use those hidden helper functions than to have four nearly identical tests. 

If the tests pass for one of those four configurations then they *should* pass for the other three, but I thought it better to test it and be sure than to assume.

Reply:
Pytest fixtures are awesome:
https://docs.pytest.org/en/latest/fixture.html